File: skills/tech-reports-skills/svg-diagram-verification/scripts/svg_geom_check.js
// svg_geom_check.js — 整段粘进 browser_console 的 expression 执行。
// 前提：页面已打开且含单个 <svg>。返回 JSON：outOfCanvas / overlaps / textOutOfRange。
(() => {
  const svg = document.querySelector('svg');
  if (!svg) return JSON.stringify({error: 'no svg found on page'});
  const vb = svg.viewBox.baseVal;
  const rects = [...svg.querySelectorAll('rect')]
    .filter(r => r.getAttribute('rx'))   // 组件框；背景/图案 rect 无 rx，自动排除
    .map(r => ({x:+r.getAttribute('x'), y:+r.getAttribute('y'),
                w:+r.getAttribute('width'), h:+r.getAttribute('height')}));
  // 1) 越界
  const outOfCanvas = rects.filter(b =>
    b.x < 0 || b.y < 0 || b.x + b.w > vb.width || b.y + b.h > vb.height);
  // 2) 重叠（跳过同几何的双 rect 遮罩对——architecture-diagram 的预期模式）
  const overlaps = [];
  for (let i = 0; i < rects.length; i++) for (let j = i + 1; j < rects.length; j++) {
    const a = rects[i], b = rects[j];
    if (a.x < b.x + b.w && a.x + a.w > b.x && a.y < b.y + b.h && a.y + a.h > b.y) {
      if (a.x === b.x && a.y === b.y && a.w === b.w && a.h === b.h) continue;
      overlaps.push([a, b]);
    }
  }
  // 3) 文字越界（仅查 x 锚点；软溢出需 vision）
  const texts = [...svg.querySelectorAll('text')];
  const textOutOfRange = texts.filter(t => {
    const tx = +t.getAttribute('x');
    return tx < 0 || tx > vb.width;
  }).length;
  return JSON.stringify({viewBox: [vb.width, vb.height], boxes: rects.length,
    outOfCanvas, overlaps: overlaps.length, overlapPairs: overlaps.slice(0, 5),
    textOutOfRange, totalText: texts.length});
})()

File: skills/svg-diagram-verification/scripts/svg_geom_check.py
#!/usr/bin/env python3
"""Browser-free SVG geometry check for architecture-diagram HTML output.

Usage:
    python3 svg_geom_check.py <diagram.html> [containers.json]

No browser, no vision needed. Checks:
  1. rect/text out of viewBox bounds
  2. text soft-overflow: centered text wider than its container box
     (CJK-aware width estimation; ASCII ~= 0.62em for JetBrains Mono)

containers.json is optional; shape: {"name": [x, y, w, h], ...}
Exits 0 on pass, 1 on problems.
"""
import sys, json, re, unicodedata
import xml.etree.ElementTree as ET

NS = "{http://www.w3.org/2000/svg}"


def is_cjk(ch):
    return unicodedata.east_asian_width(ch) in ("W", "F")


def text_width(s, fs):
    w = 0.0
    for ch in s:
        w += fs if is_cjk(ch) else 0.62 * fs
    return w


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(2)
    html = open(sys.argv[1], encoding="utf-8").read()
    m = re.search(r"<svg.*?</svg>", html, re.S)
    if not m:
        print("no <svg> found")
        sys.exit(1)
    # <svg> already carries xmlns; do NOT re-inject (duplicate-attribute error)
    root = ET.fromstring(m.group(0))

    vb = list(map(float, root.attrib.get("viewBox", "0 0 1200 760").split()))
    VW, VH = vb[2], vb[3]
    print(f"viewBox: {vb[0]} {vb[1]} {VW} {VH}")

    containers = {}
    if len(sys.argv) >= 3:
        containers = json.load(open(sys.argv[2], encoding="utf-8"))

    problems = []
    for el in root.iter():
        tag = el.tag.replace(NS, "")
        if tag == "rect":
            x = float(el.attrib.get("x", 0)); y = float(el.attrib.get("y", 0))
            w = float(el.attrib.get("width", 0)); h = float(el.attrib.get("height", 0))
            if x < 0 or y < 0 or x + w > VW or y + h > VH:
                problems.append(f"rect out of bounds: x={x} y={y} w={w} h={h}")
        elif tag == "text":
            x = float(el.attrib.get("x", 0)); y = float(el.attrib.get("y", 0))
            fs = float(el.attrib.get("font-size", 10))
            anchor = el.attrib.get("text-anchor", "start")
            s = "".join(el.itertext()).strip()
            tw = text_width(s, fs)
            left = x - tw / 2 if anchor == "middle" else x
            right = x + tw / 2 if anchor == "middle" else x + tw
            if left < 0 or right > VW or y > VH:
                problems.append(f"text out of bounds: '{s[:20]}' x={x} anchor={anchor} w~={tw:.0f} y={y}")
            if anchor == "middle":
                for name, (cx, cy, cw, ch) in containers.items():
                    if cx <= x <= cx + cw and cy <= y <= cy + ch and tw > cw:
                        problems.append(f"text overflows [{name}]: '{s[:30]}' w~={tw:.0f} > {cw}")

    if problems:
        print(f"\n!! {len(problems)} problem(s):")
        for p in problems:
            print(" -", p)
        sys.exit(1)
    print("\nOK: no out-of-bounds or soft-overflow issues")


if __name__ == "__main__":
    main()
